File: binaryTreePreorderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class stack:
    def __init__(self):
        self.data = []
        
    def push(self, item):
        self.data.insert(0, item)
        
    def getTop(self):
        return self.data[0]
        
    def pop(self):
        result = self.data[0]
        self.data=self.data[1:]
        return result
        
    def isEmpty(self):
        return len(self.data) == 0
        
class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        result = []
        s = stack()
        s.push(root)
        while root and not s.isEmpty():
            cur = s.pop()
            result.append(cur.val)
            if cur.right:
                s.push(cur.right)
            if cur.left:
                s.push(cur.left)
        return result

File: JAVA/BinaryTreePreorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        Stack<TreeNode> stk = new Stack<TreeNode>();
        if (root == null) return result;
        stk.push(root);
        TreeNode temp;
        while (!stk.empty()){
            temp = stk.pop();
            result.add(temp.val);
            if (temp.right != null) stk.push(temp.right);
            if (temp.left != null) stk.push(temp.left);
        }
        return result;
    }
}